File: clients/python/net/host.py
import socket, threading
from net.protocol import encode, decode, MsgType


class Host:
    def __init__(self, port, game, session_name):
        self.port = port
        self.game = game
        self.session_name = session_name
        self.clients = {}   # player_id -> socket
        self._lock = threading.Lock()
        self._server = None
        self._running = False

    def start(self):
        self._server = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        self._server.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        self._server.bind(('', self.port))
        self._server.listen(8)
        self._server.settimeout(1.0)
        self._running = True
        threading.Thread(target=self._accept_loop, daemon=True).start()

    def stop(self):
        self._running = False
        if self._server:
            self._server.close()

    def _accept_loop(self):
        while self._running:
            try:
                conn, _ = self._server.accept()
                threading.Thread(target=self._handle, args=(conn,), daemon=True).start()
            except socket.timeout:
                continue
            except OSError:
                break

    def _handle(self, conn):
        player_id = None
        try:
            for line in conn.makefile('rb'):
                if not line.strip():
                    continue
                try:
                    msg = decode(line)
                except Exception:
                    continue
                t = msg.get('type')
                if t == MsgType.JOIN:
                    player_id = msg['from']
                    with self._lock:
                        self.clients[player_id] = conn
                    self.broadcast({'type': MsgType.PLAYER_LIST,
                                    'players': list(self.clients.keys())})
                elif t == MsgType.MOVE:
                    self._handle_move(msg)
                elif t == MsgType.CHAT:
                    self.broadcast(msg)
                elif t == MsgType.LEAVE:
                    break
        finally:
            if player_id:
                with self._lock:
                    self.clients.pop(player_id, None)
                self.broadcast({'type': MsgType.PLAYER_LIST,
                                'players': list(self.clients.keys())})
            try:
                conn.close()
            except Exception:
                pass

    def _handle_move(self, msg):
        player_id = msg['from']
        move_data = msg.get('data', {})
        if self.game.current_turn() != player_id:
            self._send_to(player_id, {'type': MsgType.ERROR, 'message': 'not your turn'})
            return
        if not self.game.validate_move(player_id, move_data):
            self._send_to(player_id, {'type': MsgType.ERROR, 'message': 'invalid move'})
            return
        self.game.apply_move(player_id, move_data)
        with self._lock:
            pids = list(self.clients.keys())
        for pid in pids:
            state_msg = {'type': MsgType.STATE,
                         'data': self.game.get_state(perspective=pid)}
            self._send_to(pid, state_msg)
        done, winner = self.game.is_over()
        if done:
            self.broadcast({'type': MsgType.GAME_OVER, 'winner': winner})

    def broadcast(self, msg):
        data = encode(msg)
        with self._lock:
            conns = list(self.clients.values())
        for c in conns:
            self._send_raw(c, data)

    def _send_to(self, player_id, msg):
        with self._lock:
            conn = self.clients.get(player_id)
        if conn:
            self._send_raw(conn, encode(msg))

    def _send_raw(self, conn, data):
        try:
            conn.sendall(data)
        except Exception:
            pass

File: clients/node/src/net/host.js
'use strict';
const net = require('net');
const { encode, decode, MsgType } = require('./protocol');

class Host {
  constructor({ port, gameName = null, maxPlayers = 2 }) {
    this.port = port;
    this._gameName = gameName;
    this._maxPlayers = maxPlayers;
    this._clients = new Map();   // playerId -> socket
    this._server = null;
    this._running = false;
    this.game = null;
  }

  start() {
    this._server = net.createServer(sock => this._handleClient(sock));
    this._running = true;
    return new Promise((resolve, reject) => {
      this._server.once('error', reject);
      this._server.once('listening', () => { this._server.removeListener('error', reject); resolve(); });
      this._server.listen(this.port);
    });
  }

  stop() {
    this._running = false;
    if (this._server) this._server.close();
    for (const sock of this._clients.values()) {
      try { sock.destroy(); } catch (_) {}
    }
  }

  _handleClient(sock) {
    let buf = '';
    let playerId = null;

    sock.on('data', chunk => {
      buf += chunk.toString('utf8');
      const lines = buf.split('\n');
      buf = lines.pop();
      for (const line of lines) {
        if (!line.trim()) continue;
        let msg;
        try { msg = decode(line); } catch (_) { continue; }
        this._handleMessage(sock, msg, (pid) => { playerId = pid; });
      }
    });

    sock.on('close', () => {
      if (playerId) {
        this._clients.delete(playerId);
        this.broadcast({ type: MsgType.PLAYER_LIST, players: [...this._clients.keys()] });
      }
    });

    sock.on('error', () => {});
  }

  _handleMessage(sock, msg, setId) {
    const type = msg.type;
    if (type === MsgType.JOIN) {
      const pid = msg.from;
      setId(pid);
      this._clients.set(pid, sock);
      const players = [...this._clients.keys()];
      this.broadcast({ type: MsgType.PLAYER_LIST, players });
      if (!this.game && this._gameName && players.length >= this._maxPlayers) {
        try {
          const { loadGameClasses } = require('../lobby/session');
          const classes = loadGameClasses();
          if (classes[this._gameName]) {
            this.game = new classes[this._gameName]();
            this.game.start(players);
            this.broadcast({ type: MsgType.GAME_START, game: this._gameName, players });
          }
        } catch (e) {
          process.stderr.write(`[host] game start error: ${e.message}\n`);
        }
      }
    } else if (type === MsgType.MOVE) {
      this._handleMove(msg);
    } else if (type === MsgType.CHAT) {
      this.broadcast(msg);
    } else if (type === MsgType.LEAVE) {
      sock.destroy();
    }
  }

  _handleMove(msg) {
    const { t } = require('../i18n');
    const pid = msg.from;
    if (!this.game) return;
    if (this.game.currentTurn() !== pid) {
      this._sendTo(pid, { type: MsgType.ERROR, message: t('error.not_your_turn') });
      return;
    }
    if (!this.game.validateMove(pid, msg.data || {})) {
      this._sendTo(pid, { type: MsgType.ERROR, message: t('error.invalid_move') });
      return;
    }
    this.game.applyMove(pid, msg.data || {});
    for (const p of this._clients.keys()) {
      this._sendTo(p, { type: MsgType.STATE, data: this.game.getState(p) });
    }
    const [done, winner] = this.game.isOver();
    if (done) {
      this.broadcast({ type: MsgType.GAME_OVER, winner });
    }
  }

  broadcast(msg) {
    const data = encode(msg);
    for (const sock of this._clients.values()) {
      try { sock.write(data); } catch (_) {}
    }
  }

  _sendTo(playerId, msg) {
    const sock = this._clients.get(playerId);
    if (sock) {
      try { sock.write(encode(msg)); } catch (_) {}
    }
  }
}

module.exports = { Host };
